### PR TITLE
CM-1102: How to Prevent a Hangfire Job From Running When it is Already Active

### PR DESCRIPTION
**Pull request by sswietoniowski**
Provided an example solution based on the following:

https://timdeschryver.dev/blog/prevent-a-hangfire-job-from-running-when-it-is-already-active

Also used:

https://code-maze.com/hangfire-with-asp-net-core/
https://code-maze.com/aspnetcore-different-ways-to-run-background-tasks/
https://code-maze.com/chsarp-the-differences-between-quartz-net-and-hangfire/

https://codewithstyle.info/real-bug-story-long-running-jobs-hangfire/
https://discuss.hangfire.io/t/how-do-i-prevent-creation-of-duplicate-jobs/1222/5
https://gist.github.com/odinserj/a8332a3f486773baa009

### CODE REVIEW

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs [hunk 1]
```diff
@@ -0,0 +1,70 @@
...
+        var jobId = _backgroundJobClient.Enqueue<JobService>(jobService => jobService.RunJob1Async());
+        
+        _logger.LogInformation("Created job '{JobName}' with ID '{JobId}'", Job1, jobId);
+        
+        return NoContent();
+    }
+    
+    [HttpPost("create-job-2")]
+    [ProducesResponseType(StatusCodes.Status204NoContent)]
+    [ProducesResponseType(StatusCodes.Status400BadRequest)]
```
Comment (sahebagarwal-codemaze) [inline aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs:52]: Looks like StatusCodes.Status400BadRequest is not required

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs [hunk 2]
```diff
@@ -0,0 +1,40 @@
...
+
+    private readonly HttpClient _client;
+
+    public Job2IntegrationTests(ApiApplicationFactory factory)
+    {
+        _client = factory.CreateClient();
+    }
+
+    [Fact]
+    public async Task GivenJobsController_WhenCreatingManyOccurrencesOfAGivenJobButUsingSkipConcurrentExecutionAttribute_ThenTheExtraJobsAreBeingSkipped()
```
Comment (sahebagarwal-codemaze) [inline aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs:15]: Function name is too big, try to shorten it

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Services/JobService.cs [hunk 3]
```diff
@@ -0,0 +1,29 @@
...
+    private const int OperationDurationInSeconds = 10 * 60;
+
+    private readonly ILogger<JobService> _logger;
+
+    public JobService(ILogger<JobService> logger)
+    {
+        _logger = logger ?? throw new ArgumentNullException(nameof(logger));
+    }
+
+    public async Task RunJob1Async() => await PerformLongRunningOperationAsync(nameof(RunJob1Async));
```
Comment (sahebagarwal-codemaze) [inline aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Services/JobService.cs:14]: 1. Add DisableConcurrentExecution attribute for RunJob1Async
2. Also add test case for this job as well

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs (hunk 1 above)
Comment (sswietoniowski) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs]: Done

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs (hunk 2 above)
Comment (sswietoniowski) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs]: Done

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Services/JobService.cs (hunk 3 above)
Comment (sswietoniowski) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Services/JobService.cs:14]: While I can add this attribute, it is not working (at least while we're using SqlLite); a test was added, proving that this attribute is not working as expected. BTW it is not only my experience: https://github.com/HangfireIO/Hangfire/issues/1960

Comment (sswietoniowski) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Services/JobService.cs:14]: One more thing: there is no way to test [DisableConcurrentExecution] with the in-memory storage. Everything works if we're using Microsoft SQL Server, but not for the InMemoryStorage nor the SqlLite provider:

    //config.UseInMemoryStorage();
    config.UseSqlServerStorage(builder.Configuration.GetConnectionString("HangfireConnection"),
        new SqlServerStorageOptions
        {
            PrepareSchemaIfNecessary = true,
        });
So, while I can test that this attribute is not working for InMemoryStorage, there is no easy way to test a working scenario with Microsoft SQL Server, which is the main reason why I skipped this test.

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs (hunk 1 above)
Comment (sahebagarwal-codemaze) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/HowToPreventAHangfireJobFromRunning/Controllers/JobsController.cs]: Reviewed

File: aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs (hunk 2 above)
Comment (sahebagarwal-codemaze) [reply aspnetcore-features/HowToPreventAHangfireJobFromRunning/Tests/Job2IntegrationTest.cs]: Reviewed

**Pull request merged**